### DISCUSSION TITLE
SHAP base value isn't what I would expect

### DISCUSSION THREAD

**Discussion by HanLum** [shap/shap#3994] [Q&A]
I've relentlessly tried to find the answer to this for the past 3 days but find myself going round in circles... 

I'm using XGBoost for binary classification and I have a highly imbalanced dataset of 153 patients (25/16.3% positive class). Due to this imbalance, I've weighted the dataset in the model specification using the recommendation of negative class/positive class (=5.12). I noticed that the base value in my SHAP waterfall plot isn't what I would expect and found information suggesting it's probably to do with the weighting. However, no matter what I try, I can't calculate the base value given in the waterfall plot. 

I've found conflicting information suggesting the base value is both the average probability after fitting the model to the training data and also that it's the average of the outcome variable in the training data (the proportion of positive cases in the case of classification). In a different model where I didn't need to weight the sample, I've managed to make my calculation match the base value presuming it's the log odds of the proportion of positive cases.

Here is the maths for my imbalanced model: 
Proportion of positive cases = .163
Base value=log(.163/(1-.163))=-1.636074

but the base value in my SHAP plot is .212. 

I've tried averaging the post-hoc predictions, averaging the predicted class based on a threshold for the predictions and weighted versions of both of those (none of which made much sense as a base value and considering weights should be accounted for already)

Is there a way to explain this discrepancy? 

Best wishes,
Hannah 

Edit: 
I've realised the base value is the average of the predictions in log odds but it still doesn't make sense!

My predictions for my training set are:

mean(log(Prediction/(1-Prediction)))=-0.4114967

weight=(Diagnosis=="LVO")*5.12+(Diagnosis!="LVO")
weighted.mean(Prediction, weight)
0.5035255
weighted.mean(log(Prediction/(1-Prediction)),weight)
0.01210124

Do you know where I’m going wrong?


Prediction	Diagnosis
0.3052258	non-LVO
0.5903241	LVO
0.2187342	non-LVO
0.7869894	non-LVO
0.5131153	non-LVO
0.2442923	non-LVO
0.3902267	non-LVO
0.4408666	non-LVO
0.3146603	non-LVO
0.6517534	non-LVO
0.1657728	non-LVO
0.2741036	non-LVO
0.5972106	non-LVO
0.2800872	non-LVO
0.4735073	non-LVO
0.2150321	non-LVO
0.3105696	non-LVO
0.6260788	LVO
0.4405199	non-LVO
0.2365895	non-LVO
0.2173986	non-LVO
0.5208161	non-LVO
0.2450289	non-LVO
0.2592905	non-LVO
0.6008621	non-LVO
0.317544	non-LVO
0.3666764	non-LVO
0.3969755	non-LVO
0.6935067	LVO
0.3690063	non-LVO
0.3367071	non-LVO
0.5485341	non-LVO
0.2035597	non-LVO
0.6554747	non-LVO
0.5199919	LVO
0.240981	non-LVO
0.5392051	non-LVO
0.2225941	non-LVO
0.5447863	LVO
0.2310689	non-LVO
0.4328837	non-LVO
0.2444467	non-LVO
0.7200866	LVO
0.355975	non-LVO
0.3985373	non-LVO
0.3906049	LVO
0.2924811	non-LVO
0.172852	non-LVO
0.1688925	non-LVO
0.3979591	non-LVO
0.4834989	non-LVO
0.3155866	non-LVO
0.1594569	non-LVO
0.5721162	non-LVO
0.1815405	non-LVO
0.2939298	non-LVO
0.4930511	LVO
0.4315915	non-LVO
0.1815248	non-LVO
0.5370129	LVO
0.2814984	non-LVO
0.7545432	LVO
0.4604395	non-LVO
0.6929719	non-LVO
0.1758476	non-LVO
0.5214938	non-LVO
0.3523999	non-LVO
0.4828569	non-LVO
0.1685236	non-LVO
0.8284138	LVO
0.2140754	non-LVO
0.8149475	LVO
0.560429	non-LVO
0.2869403	non-LVO
0.2789971	non-LVO
0.4808701	non-LVO
0.4894908	non-LVO
0.7047694	non-LVO
0.7613413	LVO
0.2655301	non-LVO
0.6385984	LVO
0.3201781	non-LVO
0.7673382	LVO
0.73831	LVO
0.6528715	LVO
0.4756911	non-LVO
0.219945	non-LVO
0.1840003	non-LVO
0.2322547	non-LVO
0.4571077	non-LVO
0.7159563	LVO
0.1977195	non-LVO
0.1940509	non-LVO
0.2430587	non-LVO
0.4110004	non-LVO
0.3367475	non-LVO
0.2391446	non-LVO
0.5063161	non-LVO
0.2076635	non-LVO
0.3998126	non-LVO
0.4336176	non-LVO
0.442129	non-LVO
0.6303618	LVO
0.3031145	non-LVO
0.4755053	non-LVO
0.3836403	non-LVO
0.6912513	non-LVO
0.4606192	LVO
0.6196795	LVO
0.4283525	non-LVO
0.2026801	non-LVO
0.5110471	non-LVO
0.2280304	non-LVO
0.5218878	non-LVO
0.4894836	non-LVO
0.3269886	non-LVO
0.3866788	non-LVO
0.2433265	non-LVO
0.7871559	LVO
0.2575517	non-LVO
0.1921343	non-LVO
0.4225675	non-LVO
0.3512519	non-LVO
0.3264217	non-LVO
0.2255149	non-LVO
0.7954208	LVO
0.4433914	non-LVO
0.5970163	non-LVO
0.3177105	non-LVO
0.2718393	non-LVO
0.3250923	non-LVO
0.217085	non-LVO
0.1555425	non-LVO
0.5274982	non-LVO
0.4504259	non-LVO
0.4663254	LVO
0.6237264	non-LVO
0.5898137	non-LVO
0.2670143	non-LVO
0.4392689	non-LVO
0.2318519	non-LVO
0.5901781	LVO
0.1721892	non-LVO
0.4234818	non-LVO
0.2837489	non-LVO
0.4887586	non-LVO
0.258081	non-LVO
0.4306361	non-LVO
0.3149272	non-LVO
0.3103408	non-LVO
0.3983943	non-LVO
0.1811696	non-LVO
0.4602375	non-LVO

**Comment by HanLum** [comment]
@slundberg please save me

**Comment by linminhtoo** [comment]
have you tried digging into the source code? there should be a section where SHAP calculates the base value. in KernelExplainer I believe this is called `model_null` 

do you have code on how you're setting up the Explainer class & putting your trained XGB through it to get the SHAP values & waterfall plots? in my experience the value of the parameter `link` as well as the nature of what your XGBoost actually returns (e.g. does it return logits or probas) would make a big difference as to what SHAP "sees" or does with the predictions
